### PR TITLE
Add MPS implementation of CTC Loss based on CUDA version

### LINKED ISSUE #153976: MPS (Metal) CTC Loss Implementation (Gist link)
Hello PyTorch team,

I've created a Metal (MPS) implementation of the CTC Loss based on the CUDA version written by Thomas Viehmann, for use on Apple Silicon.

It is available here as a public Gist:  
https://gist.github.com/aspinto/88a337006defcf3d3af113cbc15d48c8

This might help others who want to run CTC Loss training using MPS backend natively.

Best regards,  
André

### PR DESCRIPTION
**Pull request by aspinto**
Fixes #153976
This PR adds a Metal (MPS) backend implementation of `ctc_loss`, based on the existing CUDA version.

### Changes:
- Introduced `ctc_loss_mps` and `ctc_loss_backward_mps` in native code.
- Added `.metal` shader file and Objective-C++ wrapper (`LossCTC.mm`).
- Updated `native_functions.yaml` to register the MPS dispatch.
- Extended `TensorBody.h` and Python bindings to support `.mps()` conversion.
- Minor VSCode config to enable Metal syntax highlighting.

### Notes:
- Implementation tested with `float32` inputs.
- Produces matching loss with CUDA version (confirmed within numerical precision).
- Gradients are consistent across both backends.

Looking forward to feedback!

### CODE REVIEW

**Comment by aspinto** [comment]
Thanks for reviewing! Let me know if anything needs adjustment. Happy to improve as needed.

**Comment by malfet** [comment]
@aspinto I'm quite surprised you don't need to adjust any tests, as we should have test in place that likely fails because op is not implemented

File: aten/src/ATen/native/mps/LossCTC.mm [hunk 1]
```diff
@@ -0,0 +1,842 @@
...
+        [encoder setBytes:&gr_strides[1]  length:sizeof(int64_t) atIndex:13];
+        [encoder setBytes:&gr_strides[2]   length:sizeof(int64_t) atIndex:14];
+
+        // buffer(15–17): strides de log_probs
+        int64_t lp_strides[] = {
+            log_probs.stride(0), // T
+            log_probs.stride(1), // B
+            log_probs.stride(2), // D
+        };
+        [encoder setBytes:&lp_strides[0] length:sizeof(int64_t) atIndex:15];
```
Comment (malfet) [inline aten/src/ATen/native/mps/LossCTC.mm:685]: Please use `mtl_setArg`/`mtl_setArgs` to do something of that nature

**Review by malfet — CHANGES_REQUESTED** [review]
- This PR does add ops to native_functions.yml (here for example https://github.com/pytorch/pytorch/pull/154044 )
- It only registers it for float data type, but it needs to consider half and bfloat16 (and may be integer types as well?)
- Please use `long` types on metal only when needed

**Comment by aspinto** [comment]
> @aspinto I'm quite surprised you don't need to adjust any tests, as we should have test in place that likely fails because op is not implemented

You're absolutely right — initially, I had local changes (including native_functions.yaml) that were not properly committed and pushed during the first submission. I’ve since fixed that, and the missing registration and related definitions should now be included in the latest commit.

Apologies for the confusion and thanks again for the review!

**Comment by codebreakymefix** [comment]
Any updates on this?

**Comment by nckcard** [comment]
> Any updates on this?

Sure hope that this gets reviewed and merged soon!

**Comment by nabil6391** [comment]
Hi will @aspinto work on this? Or what can be done to make progress on this without aspinto
